![green-divider](https://user-images.githubusercontent.com/7065401/52071924-c003ad80-2562-11e9-8297-1c6595f8a7ff.png)
## ***Ejercicio Practico Video Games Sales***
        Poniendo en practica lo aprendido en el curso se FreeCodeCamp "Data Analysis with Python"

![green-divider](https://user-images.githubusercontent.com/7065401/52071924-c003ad80-2562-11e9-8297-1c6595f8a7ff.png)
#### 
**Dia 1 Orientación e Inspección (El Mapa del Tesoro)**

Tema del Día: Inspección de DataFrames. Repaso profundo de .info(), .describe(), .shape, .columns, .dtypes. Entender qué te dice cada uno sobre la "salud" de tus datos.

Aplicación al Proyecto: Carga el CSV de videojuegos y ejecuta todas las funciones de inspección. En una celda de Markdown, anota tus hallazgos iniciales: ¿Cuántas filas/columnas hay? ¿Qué tipos de datos ves? ¿Hay valores nulos a simple vista?

In [3]:
import pandas as pd
import numpy as np

print(pd.__version__)

2.3.1


In [4]:
df_videogames = pd.read_csv('vgsales.csv')

df_videogames.head()

Rank                      Name Platform    Year         Genre Publisher  \
0     1                Wii Sports      Wii  2006.0        Sports  Nintendo   
1     2         Super Mario Bros.      NES  1985.0      Platform  Nintendo   
2     3            Mario Kart Wii      Wii  2008.0        Racing  Nintendo   
3     4         Wii Sports Resort      Wii  2009.0        Sports  Nintendo   
4     5  Pokemon Red/Pokemon Blue       GB  1996.0  Role-Playing  Nintendo   

   NA_Sales  EU_Sales  JP_Sales  Other_Sales  Global_Sales  
0     41.49     29.02      3.77         8.46         82.74  
1     29.08      3.58      6.81         0.77         40.24  
2     15.85     12.88      3.79         3.31         35.82  
3     15.75     11.01      3.28         2.96         33.00  
4     11.27      8.89     10.22         1.00         31.37

In [5]:
df_videogames.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16598 entries, 0 to 16597
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Rank          16598 non-null  int64  
 1   Name          16598 non-null  object 
 2   Platform      16598 non-null  object 
 3   Year          16327 non-null  float64
 4   Genre         16598 non-null  object 
 5   Publisher     16540 non-null  object 
 6   NA_Sales      16598 non-null  float64
 7   EU_Sales      16598 non-null  float64
 8   JP_Sales      16598 non-null  float64
 9   Other_Sales   16598 non-null  float64
 10  Global_Sales  16598 non-null  float64
dtypes: float64(6), int64(1), object(4)
memory usage: 1.4+ MB


![green-divider](https://user-images.githubusercontent.com/7065401/52071924-c003ad80-2562-11e9-8297-1c6595f8a7ff.png)
###  Hallazgos de inspección .info()

-  Me dice que para la celdas Year, cuenta con datos NaN, por lo que necesitare una estrategia para limpiarlos y quiza por eso Panda lo toma como un dato float64 cuando deberia ser int64.
-  Me dice que para la celdas Publisher, cuenta con datos NaN, por lo que necesitare una estrategia para limpiarlos

In [6]:
df_videogames.describe()

Rank          Year      NA_Sales      EU_Sales      JP_Sales  \
count  16598.000000  16327.000000  16598.000000  16598.000000  16598.000000   
mean    8300.605254   2006.406443      0.264667      0.146652      0.077782   
std     4791.853933      5.828981      0.816683      0.505351      0.309291   
min        1.000000   1980.000000      0.000000      0.000000      0.000000   
25%     4151.250000   2003.000000      0.000000      0.000000      0.000000   
50%     8300.500000   2007.000000      0.080000      0.020000      0.000000   
75%    12449.750000   2010.000000      0.240000      0.110000      0.040000   
max    16600.000000   2020.000000     41.490000     29.020000     10.220000   

        Other_Sales  Global_Sales  
count  16598.000000  16598.000000  
mean       0.048063      0.537441  
std        0.188588      1.555028  
min        0.000000      0.010000  
25%        0.000000      0.060000  
50%        0.010000      0.170000  
75%        0.040000      0.470000  
max       10.570000     82.740000

![green-divider](https://user-images.githubusercontent.com/7065401/52071924-c003ad80-2562-11e9-8297-1c6595f8a7ff.png)
###  Hallazgos de inspección .describe()

-  Me dice que para la celda Global_Sales, cuenta con una desviasion estandar de 1.555028 se presume alta, lo que sugiere presencia de outliers (juegos con ventas masivas que podrian sesgar el promedio)

In [7]:
df_videogames.shape

(16598, 11)

In [8]:
df_videogames.columns

Index(['Rank', 'Name', 'Platform', 'Year', 'Genre', 'Publisher', 'NA_Sales',
       'EU_Sales', 'JP_Sales', 'Other_Sales', 'Global_Sales'],
      dtype='object')

In [9]:
df_videogames.dtypes

Rank              int64
Name             object
Platform         object
Year            float64
Genre            object
Publisher        object
NA_Sales        float64
EU_Sales        float64
JP_Sales        float64
Other_Sales     float64
Global_Sales    float64
dtype: object

![green-divider](https://user-images.githubusercontent.com/7065401/52071924-c003ad80-2562-11e9-8297-1c6595f8a7ff.png)
#### Practica del dia 2

In [10]:
df_videogames['Name'].loc[5:11]

5                        Tetris
6         New Super Mario Bros.
7                      Wii Play
8     New Super Mario Bros. Wii
9                     Duck Hunt
10                   Nintendogs
11                Mario Kart DS
Name: Name, dtype: object

In [11]:
df_videogames[['Name','Platform','Global_Sales']].loc[5:10]

Name Platform  Global_Sales
5                      Tetris       GB         30.26
6       New Super Mario Bros.       DS         30.01
7                    Wii Play      Wii         29.02
8   New Super Mario Bros. Wii      Wii         28.62
9                   Duck Hunt      NES         28.31
10                 Nintendogs       DS         24.76

In [12]:
df_videogames.iloc[0]

Rank                     1
Name            Wii Sports
Platform               Wii
Year                2006.0
Genre               Sports
Publisher         Nintendo
NA_Sales             41.49
EU_Sales             29.02
JP_Sales              3.77
Other_Sales           8.46
Global_Sales         82.74
Name: 0, dtype: object

In [13]:
df_videogames.tail()

Rank                                              Name Platform  \
16593  16596                Woody Woodpecker in Crazy Castle 5      GBA   
16594  16597                     Men in Black II: Alien Escape       GC   
16595  16598  SCORE International Baja 1000: The Official Game      PS2   
16596  16599                                        Know How 2       DS   
16597  16600                                  Spirits & Spells      GBA   

         Year     Genre   Publisher  NA_Sales  EU_Sales  JP_Sales  \
16593  2002.0  Platform       Kemco      0.01      0.00       0.0   
16594  2003.0   Shooter  Infogrames      0.01      0.00       0.0   
16595  2008.0    Racing  Activision      0.00      0.00       0.0   
16596  2010.0    Puzzle    7G//AMES      0.00      0.01       0.0   
16597  2003.0  Platform     Wanadoo      0.01      0.00       0.0   

       Other_Sales  Global_Sales  
16593          0.0          0.01  
16594          0.0          0.01  
16595          0.0          0.01  
16596          0.0          0.01  
16597          0.0          0.01

In [14]:
df_videogames[['Name','Platform','Year']].iloc[16593:16598]

Name Platform    Year
16593                Woody Woodpecker in Crazy Castle 5      GBA  2002.0
16594                     Men in Black II: Alien Escape       GC  2003.0
16595  SCORE International Baja 1000: The Official Game      PS2  2008.0
16596                                        Know How 2       DS  2010.0
16597                                  Spirits & Spells      GBA  2003.0

In [15]:
df_videogames.loc[df_videogames['Genre'] == 'Sports']

Rank                           Name Platform    Year   Genre  \
0          1                     Wii Sports      Wii  2006.0  Sports   
3          4              Wii Sports Resort      Wii  2009.0  Sports   
13        14                        Wii Fit      Wii  2007.0  Sports   
14        15                   Wii Fit Plus      Wii  2009.0  Sports   
77        78                        FIFA 16      PS4  2015.0  Sports   
...      ...                            ...      ...     ...     ...   
16576  16579              Rugby Challenge 3     XOne  2016.0  Sports   
16578  16581  Outdoors Unleashed: Africa 3D      3DS  2011.0  Sports   
16579  16582              PGA European Tour      N64  2000.0  Sports   
16581  16584                      Fit & Fun      Wii  2011.0  Sports   
16587  16590       Mezase!! Tsuri Master DS       DS  2009.0  Sports   

                  Publisher  NA_Sales  EU_Sales  JP_Sales  Other_Sales  \
0                  Nintendo     41.49     29.02      3.77         8.46   
3                  Nintendo     15.75     11.01      3.28         2.96   
13                 Nintendo      8.94      8.03      3.60         2.15   
14                 Nintendo      9.09      8.59      2.53         1.79   
77          Electronic Arts      1.11      6.06      0.06         1.26   
...                     ...       ...       ...       ...          ...   
16576  Alternative Software      0.00      0.01      0.00         0.00   
16578               Mastiff      0.01      0.00      0.00         0.00   
16579            Infogrames      0.01      0.00      0.00         0.00   
16581               Unknown      0.00      0.01      0.00         0.00   
16587           Hudson Soft      0.00      0.00      0.01         0.00   

       Global_Sales  
0             82.74  
3             33.00  
13            22.72  
14            22.00  
77             8.49  
...             ...  
16576          0.01  
16578          0.01  
16579          0.01  
16581          0.01  
16587          0.01  

[2346 rows x 11 columns]

![green-divider](https://user-images.githubusercontent.com/7065401/52071924-c003ad80-2562-11e9-8297-1c6595f8a7ff.png)
#### 
**Dia 2** - *Aplicacion al proyecto*

Tema del Día: Entender a fondo la diferencia y el poder de .loc[], .iloc[] y el filtrado booleano.

Tarea: En celdas nuevas, escribe el código necesario para resolver los siguientes desafíos:

  1.  Crea un nuevo DataFrame llamado juegos_recientes que contenga solo los juegos lanzados después del año 2010.

  2.  Crea un nuevo DataFrame llamado juegos_wii que contenga todos los juegos de la plataforma "Wii".

  3.  Crea un DataFrame final llamado hits_de_nintendo que cumpla dos condiciones a la vez:

        -  Que su Publisher sea "Nintendo". Y que sus ventas en Norteamérica (NA_Sales) sean mayores a 10 millones.

In [16]:
juegos_recientes = df_videogames[df_videogames['Year'] >= 2010]

juegos_recientes

Rank                                               Name Platform  \
15        16                                 Kinect Adventures!     X360   
16        17                                 Grand Theft Auto V      PS3   
23        24                                 Grand Theft Auto V     X360   
26        27                        Pokemon Black/Pokemon White       DS   
29        30                     Call of Duty: Modern Warfare 3     X360   
...      ...                                                ...      ...   
16578  16581                      Outdoors Unleashed: Africa 3D      3DS   
16581  16584                                          Fit & Fun      Wii   
16585  16588                                             Breach       PC   
16589  16592  Chou Ezaru wa Akai Hana: Koi wa Tsuki ni Shiru...      PSV   
16596  16599                                         Know How 2       DS   

         Year         Genre               Publisher  NA_Sales  EU_Sales  \
15     2010.0          Misc  Microsoft Game Studios     14.97      4.94   
16     2013.0        Action    Take-Two Interactive      7.01      9.27   
23     2013.0        Action    Take-Two Interactive      9.63      5.31   
26     2010.0  Role-Playing                Nintendo      5.57      3.28   
29     2011.0       Shooter              Activision      9.03      4.28   
...       ...           ...                     ...       ...       ...   
16578  2011.0        Sports                 Mastiff      0.01      0.00   
16581  2011.0        Sports                 Unknown      0.00      0.01   
16585  2011.0       Shooter               Destineer      0.01      0.00   
16589  2016.0        Action         dramatic create      0.00      0.00   
16596  2010.0        Puzzle                7G//AMES      0.00      0.01   

       JP_Sales  Other_Sales  Global_Sales  
15         0.24         1.67         21.82  
16         0.97         4.14         21.40  
23         0.06         1.38         16.38  
26         5.65         0.82         15.32  
29         0.13         1.32         14.76  
...         ...          ...           ...  
16578      0.00         0.00          0.01  
16581      0.00         0.00          0.01  
16585      0.00         0.00          0.01  
16589      0.01         0.00          0.01  
16596      0.00         0.00          0.01  

[5145 rows x 11 columns]

In [17]:
juegos_wii = df_videogames[df_videogames['Platform'] == 'Wii']

juegos_wii

Rank                                               Name Platform  \
0          1                                         Wii Sports      Wii   
2          3                                     Mario Kart Wii      Wii   
3          4                                  Wii Sports Resort      Wii   
7          8                                           Wii Play      Wii   
8          9                          New Super Mario Bros. Wii      Wii   
...      ...                                                ...      ...   
16517  16520                       Teenage Mutant Ninja Turtles      Wii   
16552  16555           Mahou Sensei Negima!? Neo-Pactio Fight!!      Wii   
16573  16576                                Mini Desktop Racing      Wii   
16574  16577  Yattaman Wii: BikkuriDokkiri Machine de Mou Ra...      Wii   
16581  16584                                          Fit & Fun      Wii   

         Year     Genre                     Publisher  NA_Sales  EU_Sales  \
0      2006.0    Sports                      Nintendo     41.49     29.02   
2      2008.0    Racing                      Nintendo     15.85     12.88   
3      2009.0    Sports                      Nintendo     15.75     11.01   
7      2006.0      Misc                      Nintendo     14.03      9.20   
8      2009.0  Platform                      Nintendo     14.59      7.06   
...       ...       ...                           ...       ...       ...   
16517  2007.0    Action  Konami Digital Entertainment      0.00      0.01   
16552  2007.0  Fighting         Marvelous Interactive      0.00      0.00   
16573  2007.0    Racing                Popcorn Arcade      0.01      0.00   
16574  2008.0    Racing                   Takara Tomy      0.00      0.00   
16581  2011.0    Sports                       Unknown      0.00      0.01   

       JP_Sales  Other_Sales  Global_Sales  
0          3.77         8.46         82.74  
2          3.79         3.31         35.82  
3          3.28         2.96         33.00  
7          2.93         2.85         29.02  
8          4.70         2.26         28.62  
...         ...          ...           ...  
16517      0.00         0.00          0.01  
16552      0.01         0.00          0.01  
16573      0.00         0.00          0.01  
16574      0.01         0.00          0.01  
16581      0.00         0.00          0.01  

[1325 rows x 11 columns]

In [18]:
hits_de_nintendo = df_videogames[(df_videogames['Publisher'] == 'Nintendo') & (df_videogames['NA_Sales'] > 10)]

hits_de_nintendo

Rank                       Name Platform    Year         Genre Publisher  \
0      1                 Wii Sports      Wii  2006.0        Sports  Nintendo   
1      2          Super Mario Bros.      NES  1985.0      Platform  Nintendo   
2      3             Mario Kart Wii      Wii  2008.0        Racing  Nintendo   
3      4          Wii Sports Resort      Wii  2009.0        Sports  Nintendo   
4      5   Pokemon Red/Pokemon Blue       GB  1996.0  Role-Playing  Nintendo   
5      6                     Tetris       GB  1989.0        Puzzle  Nintendo   
6      7      New Super Mario Bros.       DS  2006.0      Platform  Nintendo   
7      8                   Wii Play      Wii  2006.0          Misc  Nintendo   
8      9  New Super Mario Bros. Wii      Wii  2009.0      Platform  Nintendo   
9     10                  Duck Hunt      NES  1984.0       Shooter  Nintendo   
18    19          Super Mario World     SNES  1990.0      Platform  Nintendo   
21    22           Super Mario Land       GB  1989.0      Platform  Nintendo   

    NA_Sales  EU_Sales  JP_Sales  Other_Sales  Global_Sales  
0      41.49     29.02      3.77         8.46         82.74  
1      29.08      3.58      6.81         0.77         40.24  
2      15.85     12.88      3.79         3.31         35.82  
3      15.75     11.01      3.28         2.96         33.00  
4      11.27      8.89     10.22         1.00         31.37  
5      23.20      2.26      4.22         0.58         30.26  
6      11.38      9.23      6.50         2.90         30.01  
7      14.03      9.20      2.93         2.85         29.02  
8      14.59      7.06      4.70         2.26         28.62  
9      26.93      0.63      0.28         0.47         28.31  
18     12.78      3.75      3.54         0.55         20.61  
21     10.83      2.71      4.18         0.42         18.14

![green-divider](https://user-images.githubusercontent.com/7065401/52071924-c003ad80-2562-11e9-8297-1c6595f8a7ff.png)
#### Practica del dia 3

In [19]:
df_ventas_globales_ps2 = df_videogames[df_videogames['Platform'] == 'PS2'].sort_values(by='Global_Sales', ascending=False).head(5)

df_ventas_globales_ps2


Rank                           Name Platform    Year   Genre  \
17    18  Grand Theft Auto: San Andreas      PS2  2004.0  Action   
24    25    Grand Theft Auto: Vice City      PS2  2002.0  Action   
28    29         Gran Turismo 3: A-Spec      PS2  2001.0  Racing   
38    39           Grand Theft Auto III      PS2  2001.0  Action   
47    48                 Gran Turismo 4      PS2  2004.0  Racing   

                      Publisher  NA_Sales  EU_Sales  JP_Sales  Other_Sales  \
17         Take-Two Interactive      9.43      0.40      0.41        10.57   
24         Take-Two Interactive      8.41      5.49      0.47         1.78   
28  Sony Computer Entertainment      6.85      5.09      1.87         1.16   
38         Take-Two Interactive      6.99      4.51      0.30         1.30   
47  Sony Computer Entertainment      3.01      0.01      1.10         7.53   

    Global_Sales  
17         20.81  
24         16.15  
28         14.98  
38         13.10  
47         11.66

In [20]:
df_videogames.isna().sum()

Rank              0
Name              0
Platform          0
Year            271
Genre             0
Publisher        58
NA_Sales          0
EU_Sales          0
JP_Sales          0
Other_Sales       0
Global_Sales      0
dtype: int64

In [21]:
df_videogames_limpio = df_videogames.dropna()

df_videogames_limpio.isna().sum()

df_videogames_limpio.info()

<class 'pandas.core.frame.DataFrame'>
Index: 16291 entries, 0 to 16597
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Rank          16291 non-null  int64  
 1   Name          16291 non-null  object 
 2   Platform      16291 non-null  object 
 3   Year          16291 non-null  float64
 4   Genre         16291 non-null  object 
 5   Publisher     16291 non-null  object 
 6   NA_Sales      16291 non-null  float64
 7   EU_Sales      16291 non-null  float64
 8   JP_Sales      16291 non-null  float64
 9   Other_Sales   16291 non-null  float64
 10  Global_Sales  16291 non-null  float64
dtypes: float64(6), int64(1), object(4)
memory usage: 1.5+ MB


In [22]:
df_videogames_dropnasubset = df_videogames.dropna(subset=['Year'])

df_videogames_dropnasubset.info()

<class 'pandas.core.frame.DataFrame'>
Index: 16327 entries, 0 to 16597
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Rank          16327 non-null  int64  
 1   Name          16327 non-null  object 
 2   Platform      16327 non-null  object 
 3   Year          16327 non-null  float64
 4   Genre         16327 non-null  object 
 5   Publisher     16291 non-null  object 
 6   NA_Sales      16327 non-null  float64
 7   EU_Sales      16327 non-null  float64
 8   JP_Sales      16327 non-null  float64
 9   Other_Sales   16327 non-null  float64
 10  Global_Sales  16327 non-null  float64
dtypes: float64(6), int64(1), object(4)
memory usage: 1.5+ MB


In [23]:
median_year = df_videogames['Year'].median()

print('Mediana calculada con .median():\n',{median_year})
      
print('\nMediana calculada por .describe()\n')
df_videogames.describe()

Mediana calculada con .median():
 {np.float64(2007.0)}

Mediana calculada por .describe()



Rank          Year      NA_Sales      EU_Sales      JP_Sales  \
count  16598.000000  16327.000000  16598.000000  16598.000000  16598.000000   
mean    8300.605254   2006.406443      0.264667      0.146652      0.077782   
std     4791.853933      5.828981      0.816683      0.505351      0.309291   
min        1.000000   1980.000000      0.000000      0.000000      0.000000   
25%     4151.250000   2003.000000      0.000000      0.000000      0.000000   
50%     8300.500000   2007.000000      0.080000      0.020000      0.000000   
75%    12449.750000   2010.000000      0.240000      0.110000      0.040000   
max    16600.000000   2020.000000     41.490000     29.020000     10.220000   

        Other_Sales  Global_Sales  
count  16598.000000  16598.000000  
mean       0.048063      0.537441  
std        0.188588      1.555028  
min        0.000000      0.010000  
25%        0.000000      0.060000  
50%        0.010000      0.170000  
75%        0.040000      0.470000  
max       10.570000     82.740000

In [24]:
df_videogames_fillna = df_videogames

df_videogames_fillna['Year'] = df_videogames_fillna['Year'].fillna(median_year)

df_videogames_fillna.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16598 entries, 0 to 16597
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Rank          16598 non-null  int64  
 1   Name          16598 non-null  object 
 2   Platform      16598 non-null  object 
 3   Year          16598 non-null  float64
 4   Genre         16598 non-null  object 
 5   Publisher     16540 non-null  object 
 6   NA_Sales      16598 non-null  float64
 7   EU_Sales      16598 non-null  float64
 8   JP_Sales      16598 non-null  float64
 9   Other_Sales   16598 non-null  float64
 10  Global_Sales  16598 non-null  float64
dtypes: float64(6), int64(1), object(4)
memory usage: 1.4+ MB


![green-divider](https://user-images.githubusercontent.com/7065401/52071924-c003ad80-2562-11e9-8297-1c6595f8a7ff.png)
#### 
**Dia 3** - *Aplicacion al proyecto*

Tema del Día: Repasar y entender las diferentes estrategias para manejar los valores nulos (NaN). Con el objetivo de tomar decisiones informadas como analista sobre cómo limpiar nuestro dataset y justificar esas decisiones.

Tarea:

1.  Investiga las columnas Year y Publisher, que ya sabemos que tienen valores nulos.

2.  Decide una estrategia de limpieza para CADA UNA de ellas. Por ejemplo: para Year, podrías decidir que eliminar las filas con año faltante es aceptable. Para Publisher, podrías decidir que es mejor rellenar los faltantes con la palabra "Unknown".

3.  Ejecuta el código para aplicar estas estrategias a tu DataFrame principal df_videogames.

4.  Justifica tus decisiones. En una celda de Markdown, escribe un breve párrafo explicando por qué elegiste cada estrategia.


In [25]:
df_videogames['Year'] = df_videogames['Year'].fillna(median_year)

df_videogames.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16598 entries, 0 to 16597
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Rank          16598 non-null  int64  
 1   Name          16598 non-null  object 
 2   Platform      16598 non-null  object 
 3   Year          16598 non-null  float64
 4   Genre         16598 non-null  object 
 5   Publisher     16540 non-null  object 
 6   NA_Sales      16598 non-null  float64
 7   EU_Sales      16598 non-null  float64
 8   JP_Sales      16598 non-null  float64
 9   Other_Sales   16598 non-null  float64
 10  Global_Sales  16598 non-null  float64
dtypes: float64(6), int64(1), object(4)
memory usage: 1.4+ MB


Para la columna 'Year' se considero usar la mediana (2.007) de la columna ya que se asume que el año de lanzamientos de los video juegos es un estimativo importante en el df, tambien los valores faltante representaban el 1.6% de la base de datos los que no afectaria considerablemente los resultados, tambien se consudero su influencia en las ventas globales. 

In [26]:
df_videogames['Publisher'] = df_videogames['Publisher'].fillna('Unknown')

df_videogames.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16598 entries, 0 to 16597
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Rank          16598 non-null  int64  
 1   Name          16598 non-null  object 
 2   Platform      16598 non-null  object 
 3   Year          16598 non-null  float64
 4   Genre         16598 non-null  object 
 5   Publisher     16598 non-null  object 
 6   NA_Sales      16598 non-null  float64
 7   EU_Sales      16598 non-null  float64
 8   JP_Sales      16598 non-null  float64
 9   Other_Sales   16598 non-null  float64
 10  Global_Sales  16598 non-null  float64
dtypes: float64(6), int64(1), object(4)
memory usage: 1.4+ MB


Para la columana 'Publisher' se considero reemplazar los valores faltantes con 'Unknown' (desconocido) ya que estos tambien podrian tener influencia sobre las ventas globales a pesar de no representar mas del 0.35% de la base de datos.

![green-divider](https://user-images.githubusercontent.com/7065401/52071924-c003ad80-2562-11e9-8297-1c6595f8a7ff.png)
#### 
#### Practica del Dia 4

In [27]:
df_videogames.isna().sum()

Rank            0
Name            0
Platform        0
Year            0
Genre           0
Publisher       0
NA_Sales        0
EU_Sales        0
JP_Sales        0
Other_Sales     0
Global_Sales    0
dtype: int64

In [29]:
numeros_en_series = pd.Series(['10', '20', '30'])

print(numeros_en_series)

0    10
1    20
2    30
dtype: object


In [30]:
numeros_enteros = numeros_en_series.astype(int)

print(numeros_enteros)

0    10
1    20
2    30
dtype: int64


In [31]:
ventas = pd.Series([5.5, 12.1, 8.8])

print(ventas)

0     5.5
1    12.1
2     8.8
dtype: float64


In [33]:
ventas = ventas.apply(lambda x: round(x))

print(ventas)

0     6
1    12
2     9
dtype: int64


![green-divider](https://user-images.githubusercontent.com/7065401/52071924-c003ad80-2562-11e9-8297-1c6595f8a7ff.png)
#### 
**Dia 4** - *Aplicacion al proyecto*

Objetivo: Aplicar estas dos nuevas herramientas para refinar nuestro DataFrame de videojuegos.

Tarea:

1.  Conversión de Tipo (.astype()): Después de la limpieza del Día 3, la columna Year sigue siendo de tipo float64. Conviértela a tipo int para que represente los años de forma correcta. Recuerda asignar el resultado para que el cambio sea permanente.

2.  Creación de Columna (.apply()): Usa .apply() y una función lambda en la columna Global_Sales para crear una nueva columna llamada Sales_Category. Esta columna debe contener el texto:

    **  'Éxito de ventas' si el valor en Global_Sales es mayor a 1 (millón).

    **  'Venta estándar' si el valor es 1 o menos.

3.  Verificación: Después de crear la nueva columna, usa .head() para visualizar tanto la columna Global_Sales como tu nueva columna Sales_Category y confirmar que la lógica funciona.

In [34]:
df_videogames['Year'] = df_videogames['Year'].astype(int)

df_videogames.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16598 entries, 0 to 16597
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Rank          16598 non-null  int64  
 1   Name          16598 non-null  object 
 2   Platform      16598 non-null  object 
 3   Year          16598 non-null  int64  
 4   Genre         16598 non-null  object 
 5   Publisher     16598 non-null  object 
 6   NA_Sales      16598 non-null  float64
 7   EU_Sales      16598 non-null  float64
 8   JP_Sales      16598 non-null  float64
 9   Other_Sales   16598 non-null  float64
 10  Global_Sales  16598 non-null  float64
dtypes: float64(5), int64(2), object(4)
memory usage: 1.4+ MB


In [35]:
df_videogames['Sales_Category'] = df_videogames['Global_Sales'].apply(lambda sales: 'Exito de ventas' if sales > 1
                                                                      else 'Venta estandar')
df_videogames.head()

Rank                      Name Platform  Year         Genre Publisher  \
0     1                Wii Sports      Wii  2006        Sports  Nintendo   
1     2         Super Mario Bros.      NES  1985      Platform  Nintendo   
2     3            Mario Kart Wii      Wii  2008        Racing  Nintendo   
3     4         Wii Sports Resort      Wii  2009        Sports  Nintendo   
4     5  Pokemon Red/Pokemon Blue       GB  1996  Role-Playing  Nintendo   

   NA_Sales  EU_Sales  JP_Sales  Other_Sales  Global_Sales   Sales_Category  
0     41.49     29.02      3.77         8.46         82.74  Exito de ventas  
1     29.08      3.58      6.81         0.77         40.24  Exito de ventas  
2     15.85     12.88      3.79         3.31         35.82  Exito de ventas  
3     15.75     11.01      3.28         2.96         33.00  Exito de ventas  
4     11.27      8.89     10.22         1.00         31.37  Exito de ventas

![green-divider](https://user-images.githubusercontent.com/7065401/52071924-c003ad80-2562-11e9-8297-1c6595f8a7ff.png)
#### 
#### Practica del Dia 5

In [36]:
cuenta_de_caracteres = df_videogames

cuenta_de_caracteres

Rank                                              Name Platform  Year  \
0          1                                        Wii Sports      Wii  2006   
1          2                                 Super Mario Bros.      NES  1985   
2          3                                    Mario Kart Wii      Wii  2008   
3          4                                 Wii Sports Resort      Wii  2009   
4          5                          Pokemon Red/Pokemon Blue       GB  1996   
...      ...                                               ...      ...   ...   
16593  16596                Woody Woodpecker in Crazy Castle 5      GBA  2002   
16594  16597                     Men in Black II: Alien Escape       GC  2003   
16595  16598  SCORE International Baja 1000: The Official Game      PS2  2008   
16596  16599                                        Know How 2       DS  2010   
16597  16600                                  Spirits & Spells      GBA  2003   

              Genre   Publisher  NA_Sales  EU_Sales  JP_Sales  Other_Sales  \
0            Sports    Nintendo     41.49     29.02      3.77         8.46   
1          Platform    Nintendo     29.08      3.58      6.81         0.77   
2            Racing    Nintendo     15.85     12.88      3.79         3.31   
3            Sports    Nintendo     15.75     11.01      3.28         2.96   
4      Role-Playing    Nintendo     11.27      8.89     10.22         1.00   
...             ...         ...       ...       ...       ...          ...   
16593      Platform       Kemco      0.01      0.00      0.00         0.00   
16594       Shooter  Infogrames      0.01      0.00      0.00         0.00   
16595        Racing  Activision      0.00      0.00      0.00         0.00   
16596        Puzzle    7G//AMES      0.00      0.01      0.00         0.00   
16597      Platform     Wanadoo      0.01      0.00      0.00         0.00   

       Global_Sales   Sales_Category  
0             82.74  Exito de ventas  
1             40.24  Exito de ventas  
2             35.82  Exito de ventas  
3             33.00  Exito de ventas  
4             31.37  Exito de ventas  
...             ...              ...  
16593          0.01   Venta estandar  
16594          0.01   Venta estandar  
16595          0.01   Venta estandar  
16596          0.01   Venta estandar  
16597          0.01   Venta estandar  

[16598 rows x 12 columns]

In [37]:
cuenta_de_caracteres['Name_Lenght'] = cuenta_de_caracteres['Name'].apply(lambda num_caracteres: len(num_caracteres))

cuenta_de_caracteres

Rank                                              Name Platform  Year  \
0          1                                        Wii Sports      Wii  2006   
1          2                                 Super Mario Bros.      NES  1985   
2          3                                    Mario Kart Wii      Wii  2008   
3          4                                 Wii Sports Resort      Wii  2009   
4          5                          Pokemon Red/Pokemon Blue       GB  1996   
...      ...                                               ...      ...   ...   
16593  16596                Woody Woodpecker in Crazy Castle 5      GBA  2002   
16594  16597                     Men in Black II: Alien Escape       GC  2003   
16595  16598  SCORE International Baja 1000: The Official Game      PS2  2008   
16596  16599                                        Know How 2       DS  2010   
16597  16600                                  Spirits & Spells      GBA  2003   

              Genre   Publisher  NA_Sales  EU_Sales  JP_Sales  Other_Sales  \
0            Sports    Nintendo     41.49     29.02      3.77         8.46   
1          Platform    Nintendo     29.08      3.58      6.81         0.77   
2            Racing    Nintendo     15.85     12.88      3.79         3.31   
3            Sports    Nintendo     15.75     11.01      3.28         2.96   
4      Role-Playing    Nintendo     11.27      8.89     10.22         1.00   
...             ...         ...       ...       ...       ...          ...   
16593      Platform       Kemco      0.01      0.00      0.00         0.00   
16594       Shooter  Infogrames      0.01      0.00      0.00         0.00   
16595        Racing  Activision      0.00      0.00      0.00         0.00   
16596        Puzzle    7G//AMES      0.00      0.01      0.00         0.00   
16597      Platform     Wanadoo      0.01      0.00      0.00         0.00   

       Global_Sales   Sales_Category  Name_Lenght  
0             82.74  Exito de ventas           10  
1             40.24  Exito de ventas           17  
2             35.82  Exito de ventas           14  
3             33.00  Exito de ventas           17  
4             31.37  Exito de ventas           24  
...             ...              ...          ...  
16593          0.01   Venta estandar           34  
16594          0.01   Venta estandar           29  
16595          0.01   Venta estandar           48  
16596          0.01   Venta estandar           10  
16597          0.01   Venta estandar           16  

[16598 rows x 13 columns]

In [38]:
df_videogames.duplicated().sum()

np.int64(0)

In [39]:
df_videogames.duplicated(subset= ['Name']).sum()

df_videogames['Name'].duplicated().sum() ##esto tambien da el mismo resultado.

np.int64(5105)

In [40]:
df_prueba = df_videogames.drop_duplicates()

df_prueba.shape

(16598, 13)

In [41]:
df_prueba_subset = df_videogames.drop_duplicates(subset=['Name'])

df_prueba_subset.shape

(11493, 13)

In [ ]:
df_prueba.to_csv('prueba.csv')

In [ ]:
df_prueba.to_csv('prueba.csv', index=False)

![green-divider](https://user-images.githubusercontent.com/7065401/52071924-c003ad80-2562-11e9-8297-1c6595f8a7ff.png)
#### 
**Dia 5** - *Aplicacion al proyecto*

Objetivo: Realizar la limpieza final de duplicados en nuestro DataFrame principal y guardarlo para las siguientes lecciones.

Tarea:

Investiga y Elimina Duplicados: En nuestro dataset, una fila duplicada no tiene mucho sentido. Pero un juego con el mismo Name y la misma Platform sí sería un duplicado problemático.

1.  Primero, comprueba si existen filas que sean una copia exacta de otras y, si las hay, elimínalas, sobrescribiendo tu DataFrame principal (df_videogames).

2.  A continuación, investiga si hay duplicados basándote en el subconjunto de columnas ['Name', 'Platform']. Decide una estrategia (por ejemplo, conservar la primera aparición (keep='first') y aplica la eliminación.

3.  Guarda el DataFrame Limpio: Guarda tu DataFrame final y limpio en un nuevo archivo llamado vgsales_limpio.csv. Asegúrate de usar index=False.

4.  Verificación Final: En una nueva celda, carga el archivo vgsales_limpio.csv en un nuevo DataFrame (ej: df_final_check) y usa .info() para asegurarte de que todo se guardó y cargó correctamente.

In [42]:
print('Numero de duplicados de fila:\n',{df_videogames.duplicated().sum()})

Numero de duplicados de fila:
 {np.int64(0)}


In [44]:
print('Numero de duplicados con coincidencia en Name y Platform\n', {df_videogames.duplicated(subset=['Name', 'Platform']).sum()})

Numero de duplicados con coincidencia en Name y Platform
 {np.int64(5)}


In [45]:
df_videogames = df_videogames.drop_duplicates(subset=['Name', 'Platform'])

df_videogames.info()

<class 'pandas.core.frame.DataFrame'>
Index: 16593 entries, 0 to 16597
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Rank            16593 non-null  int64  
 1   Name            16593 non-null  object 
 2   Platform        16593 non-null  object 
 3   Year            16593 non-null  int64  
 4   Genre           16593 non-null  object 
 5   Publisher       16593 non-null  object 
 6   NA_Sales        16593 non-null  float64
 7   EU_Sales        16593 non-null  float64
 8   JP_Sales        16593 non-null  float64
 9   Other_Sales     16593 non-null  float64
 10  Global_Sales    16593 non-null  float64
 11  Sales_Category  16593 non-null  object 
 12  Name_Lenght     16593 non-null  int64  
dtypes: float64(5), int64(3), object(5)
memory usage: 1.8+ MB


Se analizo la base de datos y se encontraron cinco (5) filas duplicas considerando si tenian la misma informacion tanto en la columna 'Name' como en la columna 'Platform', por este motivo se procedio a la eliminadion de estos datos quedando 16.593 de 16.598 registros.

In [46]:
df_videogames.to_csv('vgsales_limpio.csv', index=False)

In [47]:
df_final_check = pd.read_csv('vgsales_limpio.csv')

df_final_check.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16593 entries, 0 to 16592
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Rank            16593 non-null  int64  
 1   Name            16593 non-null  object 
 2   Platform        16593 non-null  object 
 3   Year            16593 non-null  int64  
 4   Genre           16593 non-null  object 
 5   Publisher       16593 non-null  object 
 6   NA_Sales        16593 non-null  float64
 7   EU_Sales        16593 non-null  float64
 8   JP_Sales        16593 non-null  float64
 9   Other_Sales     16593 non-null  float64
 10  Global_Sales    16593 non-null  float64
 11  Sales_Category  16593 non-null  object 
 12  Name_Lenght     16593 non-null  int64  
dtypes: float64(5), int64(3), object(5)
memory usage: 1.6+ MB


![green-divider](https://user-images.githubusercontent.com/7065401/52071924-c003ad80-2562-11e9-8297-1c6595f8a7ff.png)
#### 
#### Practica del Dia 6